napari/_qt/qt_main_window.py

```diff
@@ -1149,22 +1154,34 @@ def _screenshot(self, flash=True, canvas_only=False) -> 'QImage':
         from .utils import add_flash_animation
 
         if canvas_only:
-            img = self._qt_viewer.canvas.native.grabFramebuffer()
-            if flash:
-                add_flash_animation(self._qt_viewer._canvas_overlay)
+            canvas = self._qt_viewer.canvas
+            if size is not None:
+                prev_size = canvas.size
+                canvas.size = size
```

Comment:
I wonder there isn't a better way to implement this to achieve the desired goal. Resetting the canvas size like this is a pretty intense option for the GUI, if you were to get an error half-way through this function you might get stuck like

<img width="1328" alt="image" src="https://user-images.githubusercontent.com/6531703/158073369-394a57b6-4dae-4c27-b0a1-86e1881f39e8.png">

I would prefer if there was another way that didn't change the qt widgets.

If we do go this way we should use probably a context manager instead of a try, finally where inside the context the canvas size is changed. I believe this is the most robust and pythonic approach. what does @jni think?

Reply:
I agree.  similar request: https://github.com/napari/napari/pull/4201/commits/b22f4a79e245ee1a501c8d6762ca18f83a240204#r819771501

Reply:
I already implemented this, in the try block just below these lines. Or is that not what you meant?

Reply:
I meant literally a [context manager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) so the syntax would be something like `with custom_canavas_size(size=size):` (maybe better names are possible)

Reply:
Ah, the try finally is logically the same as the context manager.  If it's a one-off case, making a new context manager function/class is optional

Reply:
ok yeah - i don't have strong feelings here. agreed the one-off case seems like it could go either way. The more significant part of my comment is the feeling that this is not the best way to implement the desired functionality and could be very fragile when we do things like multicanvas. Can we not just take the screenshot normally and then resize using `ndi.zoom`? We faced this in napari animation and I believe used that approach

Reply:
The difference is that this approach actually renders at the requested resolution. This allows to render very big, high resolution images even if your screen is smaller than that, without sacrificing quality.

Reply:
> The difference is that this approach actually renders at the requested resolution. This allows to render very big, high resolution images even if your screen is smaller than that, without sacrificing quality.

I still can't get that to work...
(Note I'm just using `gray r` here but I'd like to be able to save images any (custom) LUT and/or labels).
I have a large mosaic merge: nearly 4k by 7k pixels.
![image](https://user-images.githubusercontent.com/76622105/158345388-fed7a463-6b64-43e8-9480-35f9e13d6732.png)

Thatt's basically the whole image. Lets say want itI then run the code visible in the console:
`screen = viewer.screenshot(size = viewer.layers[0].data.shape, canvas_only=True)` and
`viewer.add_image(screen)`
I load the output to see what it looks like:
```
screen.shape
Out[8]: (13622, 7798, 4)
```
But it's just a middle slice of the image:
![image](https://user-images.githubusercontent.com/76622105/158345705-928daaf4-a996-4916-9037-d030e8d758ab.png)

(For the record a conventional screenshot gives:
```
screen2 = viewer.screenshot(canvas_only=True)
screen2.shape
Out[12]: (1000, 1600, 4)
```
So a major loss of resolution relative to the original image.

Reply:
To clarify, @psobolewskiPhD : the canvas will be resized but will keep the zoom and centering that you are seeing. So if most of your layer is out of view, it will stay out of view when resizing. You need to set up the canvas so that you see the things you want to see, then screenshot.

The fact that your screenshot is cropped like that is probably because that's the shape of you layer, which you used as size of the canvas. So the canvas was resized to that ratio, while keeping center and zoom (therefore effectively cropping the sides of the original view).

In practice, this is the same as you resizing the napari window: it won't change what you see, just how high-res it is and the shape of the canvas.

Reply:
Hmm, I'm not sure I follow. 
You can see the shape of the layer when I took the screenshot in the first posted image. I have the zoom such that basically the whole image is shown. The 2nd posted image shows the output screenshot which is a wierd crop. To be clear that's *the output* of the `screen = viewer.screenshot(size = viewer.layers[0].data.shape, canvas_only=True)` command.

 `screen2 = viewer.screenshot(canvas_only=True)` works as expected, giving a screenshot of the entire layer, but at much lower resolution.

Reply:
So the shape of the layer is basically the same as the canvas? And it's not `(13622, 7798, 4)` or a multiple of it? And there's no scaling stuff in the way? If you pass a shape to size and get out something different (apart from the pixel doubling due to retina), then something is very wrong.

Reply:
PS: you do not need to pass `canvas_only=True`, that's the default!

Reply:
> In general, I don't see why you would pass _exactly_ the layer shape, since the result is not going to match exactly the layer unless you were pixel perfect in centering the canvas.

Gah, you're totally right. I'm stupid. I've been fixed on my original desire to save the color rendered WYSIWYG image at original resolution. But that's sort of irrelevant here & now.
So I have 2 questions and a comment.

-  When I zoom out to see my whole huge image, the representation on the canvas is downsampled, right? But when with this resize feature is that using original pixels or upsampling the downsampled image?

- Is there a `fit to window` feature to set the optimal zoom, to fill the canvas in the largest dim?

- Maybe what makes the most sense is to just provide a *scale factor*? Perhaps relative to zoom? I tested this change and interestingly it doesn't result in the pixel doubling.

Reply:
I've been fixed on my original desire to save the color rendered WYSIWYG image at original resolution.

We still should try to get that to work though!

>     * When I zoom out to see my whole huge image, the representation on the canvas is downsampled, right? But when with this resize feature is that using original pixels or upsampling the downsampled image?

The canvas renders to whatever *actual* pixel surface it's taking up on the screen, so yes. it's downsampled if it covers fewer pixels than in the data. Setting the canvas to a different size will take up that many pixels (even if bigger than the screen), so it will render as though you had a massive screen with high resolution, so no loss of resolution there (afaik!).

>     * Is there a `fit to window` feature to set the optimal zoom, to fill the canvas in the largest dim?

Not sure... `viewer.reset_view()` is the closest thing I'm familiar with.

>     * Maybe what makes the most sense is to just provide a _scale factor_? Perhaps relative to zoom? I tested this change and interestingly it doesn't result in the pixel doubling.

Not sure I follow what you tested, can you share the code you used? I agree that it might be useful to have a `zoom` parameter (mutually exclusive with size, I guess), which simply scales up the resolution of the canvas maintaining the exact same view. Probably more intuitive!

Reply:
> Not sure... `viewer.reset_view()` is the closest thing I'm familiar with.
> 

Yeah, that's close, but still leaves blank canvas bars on both top/bottm and sides of the image data.

> We still should try to get that to work though!

Sounds good to me! Just popped into my head: with scaling could one use 1/viewer.camera.zoom then crop to data.shape? I guess it requires that the view was centered, but the reset_view() does that..